framework/include/utils/GreenGaussGradient.h

```diff
@@ -0,0 +1,259 @@
+
+      if (functor.isExtrapolatedBoundaryFace(*fi))
+      {
+        if (two_term_boundary_expansion)
+        {
+          ebf_faces.push_back(fi);
+
+          // eqn. 2
+          ebf_grad_coeffs.push_back(-1. * (elem_has_info
+                                               ? (fi->faceCentroid() - fi->elemCentroid())
+                                               : (fi->faceCentroid() - fi->neighborCentroid())));
+          ebf_b.push_back(&elem_value);
+
+          // eqn. 1
+          grad_ebf_coeffs.push_back(-surface_vector);
+        }
+        else
+          // We are doing a one-term expansion for the extrapolated boundary faces, in which case we
+          // have no eqn. 2 and we have no second term in the LHS of eqn. 1. Instead we apply the
+          // element centroid value as the face value (one-term expansion) in the RHS of eqn. 1
+          grad_b += surface_vector * elem_value;
+      }
+      else
+        grad_b += surface_vector * functor(makeCDFace(*fi,
+                                                      elem_arg.correct_skewness,
+                                                      elem_arg.apply_gradient_to_skewness));
+
+      if (!volume_set)
+      {
+        // We use the FaceInfo volumes because those values have been pre-computed and cached.
+        // An explicit call to elem->volume() here would incur unnecessary expense
+        if (elem_has_info)
+        {
+          MooseMeshUtils::coordTransformFactor(
+              fi->elemCentroid(), coord, coord_type, rz_radial_coord);
+          volume = fi->elemVolume() * coord;
+        }
+        else
+        {
+          MooseMeshUtils::coordTransformFactor(
+              fi->neighborCentroid(), coord, coord_type, rz_radial_coord);
+          volume = fi->neighborVolume() * coord;
+        }
+
+        volume_set = true;
+      }
+    };
+
+    Moose::FV::loopOverElemFaceInfo(
+        *elem_arg.elem, mesh, action_functor, coord_type, rz_radial_coord);
+
+    mooseAssert(volume_set && volume > 0, "We should have set the volume");
+    grad_b /= volume;
+
+    if (coord_type == Moose::CoordinateSystemType::COORD_RZ)
+    {
+      mooseAssert(rz_radial_coord != libMesh::invalid_uint, "rz_radial_coord must be set");
+      grad_b(rz_radial_coord) -= elem_value / elem_arg.elem->vertex_average()(rz_radial_coord);
+    }
+
+    mooseAssert(
+        coord_type != Moose::CoordinateSystemType::COORD_RSPHERICAL,
+        "We have not yet implemented the correct translation from gradient to divergence for "
+        "spherical coordinates yet.");
+
+    mooseAssert(
+        ebf_faces.size() < UINT_MAX,
+        "You've created a mystical element that has more faces than can be held by unsigned "
+        "int. I applaud you.");
+    const auto num_ebfs = static_cast<unsigned int>(ebf_faces.size());
+
+    // test for simple case
+    if (num_ebfs == 0)
+      grad = grad_b;
```

Comment:
Internal face caching is handled by `getInternalFaceValue`. If we are not doing two term expansion, then we just use the element value for the face and we have never cached that. The only caching this function has ever done for face values is the two-term extrapolated boundary face values. No functionality should have have changed here with respect to caching

Reply:
I absolutely agree no functionality has changed here.

Just thinking, this is a new API to get gradients out of functors. And there's an argument to keep the face values. But if there turns out not be an extrapolated boundary, then that cache doesnt get filled 

Of course if there is no EBF, there's probably as dirichlet BC so you dont need to do that. But could be one didnt know that

Reply:
I could change the parameter name to `ebf_face_to_value_cache` to more accurately reflect what the function is doing. I'm not sure I want to try to tackle other kinds of caching within this function, at least not in this work. It would be nice to not even have to handle the ebf caching here but again that would require some rewiring